### DISCUSSION TITLE
프로그래밍에 있어서의 문서화

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#85] [프론트 기술 질문 공간]
프로그래밍에 있어서 문서화는 어느 레벨까지 필요하다고 생각하며 그 근거는 무엇입니까?

**Comment by michoball** [comment]
팀 내에서 사용하기위한 개발 문서를 만든다고 한다면 팀 문화에 대한 내용이 상세히 적혀있는 문서를 작성해야한다고 봅니다. 
PR 룰 이나, 코드 스타일 가이드, 사용하는 의존성들의 종류와 버전 등등 처음오는 개발자를 잘 온보딩하기 위해서 필요하다고 봅니다. 
프로젝트 스타일 가이드 등은 (예 : 스토리북) 디자이너나 퍼블리셔와의 협업에서도 중요히 작용할 것 같습니다. 

SI 프로젝트 같은 경우는 프로젝트 완료 후 해당 프로젝트를 유지 보수하기 위한 문서화가 필요하다고 봅니다. 
 라이브러리 버전이나 파일 구조 등  유지보수를 위한 내용 등 개발자를 위한 문서와 
 완성한 앱을 사용하는 사용자 가이드 문서가 필요하다고 봅니다. 
 
레벨로 얘기하기에는 애매하지만 여하든 세세한 부분까지 잘 작성하는 것이 좋고, 문서 또한 내용을 읽거나 찾기 쉽도록 작성하는 것이 중요하다고 생각합니다.

**Comment by WooGie911** [comment]
먼저 문서화의 정의부터 알아보겠습니다.
문서화는 튜토리얼(Tutorial), 가이드(How-to guide), 해설(Explanation), 기술 레퍼런스(Technical reference) 등 네 가지의 역할 중 하나에 속해야 하며 거기에 알맞는 구조를 가져야 합니다. 
문서는 명시적으로 이들 중 한 가지의 구조를 가져야 하고, 다른 종류의 문서들과 명확하게 구분되어야 합니다.
각각의 기능은 서로 완전히 다른 접근 방법으로 작성 되어야 하며, 유저들은 서로 다른 상황과 서로 다른 시점에 이들 문서를 찾게 되기 때문에 대부분의 소프트웨어는 저 네 가지 종류의 문서를 모두 갖출 필요가 있습니다.

- 튜토리얼
튜토리얼은 학습 위주로 이루어져 있고, 뉴비가 제품 사용을 시작할 수 있도록 해주는 강좌입니다.
ex) 어린 아이에게 요리법을 가르치는 것.

- 가이드
가이드는 목표 지향적이고 특정한 문제를 해결하는 방법을 보여주며 여러 단계로 이루어져 있습니다.
ex) 요리책에 쓰여진 조리법.

- 해설
해설은 이해를 기반으로 하며 설명하고 배경 지식과 맥락을 제공합니다.
ex) 요리의 역사에 대한 기사.

- 레퍼런스
레퍼런스는 정보를 중점으로 동작을 기술하며 정확하고 완전한 내용을 담고 있습니다.
ex) 백과사전 항목.


개인적으로 프로그래밍에 있어서 문서화는 최소화가 될수록 좋다고 생각합니다.
물론 위에서 언급한 문서화들을 상황에 맞게 사용하는 것이 가장 좋겠지만 문서화와 프로그래밍 중 하나만 선택해야 하는 상황이 온다면 프로그래밍에 더 많은 노력을 기울이는 것이 당연하다고 생각합니다.

**Comment by javascriptbiggosoo** [comment]
프로그래밍에서 문서화는 코드의 기능, 동작 방식, 사용 방법 등을 설명하는 것이며 주석,
API 문서, 개발자 가이드, 사용자 매뉴얼 등 다양한 형태를 가질 수 있습니다.
이상적이라면
본인과 동료가 시간이 지나도 이해하기 쉽도록 코드 자체가 문서화의 역할을 할 수 있
도록 직관적인 네이밍과 주석을 적절히 활용해야하며
API의 사용 방법과 입출력, 에러 처리등을 문서화해야하고
사용자가 제품을 올바르게 사용할 수 있도록 매뉴얼이나 FAQ도 마련해야합니다.
